template/DEVELOPING/database.md

```diff
@@ -0,0 +1,230 @@
+its own database it is safe to run these tests in parallel, and it is safe to
+run them in watch mode. But note that because we only spin up one database per
+test module, database state does carry over between tests in the same module.
+
+> ⚠ Test databases should be stopped automatically after each test module runs.
+> But sometimes they keep running - especially if you encounter errors during
+> a test. You might want to run `docker ps` from time to time to check whether
+> you are accumulating running databases. Test database containers are ephemeral
+> so if you stop one with `docker stop container_name` the container will be
+> removed automatically after it stops.
+
+## TypeORM Database Entities
+
+Create-originate-app sets up [TypeORM][] which encourages a code-first design which
+means that the database schema is inferred from TypeScript code. Database models
+are written as **entity** classes (classes that have an `@Entity()` annotation).
+For example, here is the [Recipe][] entity with all of the non-database-related
+annotations removed:
+
+[typeorm]: https://typeorm.io/
+[recipe]: ../packages/backend/src/recipes/models/recipe.entity.ts
+
+```ts
+@Entity()
+export class Recipe {
+  @PrimaryGeneratedColumn()
+  id!: number
+
+  @CreateDateColumn()
+  createdAt!: Date
+
+  @UpdateDateColumn()
+  updatedAt!: Date
+
+  @Column()
+  title!: string
+
+  @Column({ nullable: true, type: "text" })
+  description?: string
+
+  @ManyToMany(_type => Ingredient, ingredient => ingredient.recipes)
+  @JoinTable()
+  ingredients?: Ingredient[]
+}
+```
+
+Class variables with `Column` annotations represent database columns. There are
+multiple possible `Column` annotations including `@Column()`,
+`@PrimaryGeneratedColumn()`, `@CreateDateColumn()`, and others. We also see
+a `@JoinTable()` annotation which represents an entire table. For details see
+the [TypeORM entity documentation][]. The point is that there is sufficient
+information in the entity class to infer a full database schema for a `recipe`
+table. In combination with the `Ingredient` entity class there is sufficient
+information to infer schemas for `ingredient` and
+`recipe_ingredients_ingredient` tables too.
+
+[typeorm entity documentation]: https://typeorm.io/#/entities
+
+> ℹ️ Non-nullable columns have an exclamation mark (!) after the class variable
+> name. We expect that those variables will always have values even though we
+> don't initialize them in the class definition. The exclamation mark is
+> a non-null assertion that instructs TypeScript not to show an error.
+
+### registering database entities with the application
+
+TypeORM needs to know about your entity class when syncing the database schema.
+To make that happen every new entity class needs to be wired into a [NestJS
+module][]. For example this line from [RecipesModule][] wires in the
+`Ingredient` and `Recipe` entity classes,
+
+```ts
+imports: [TypeOrmModule.forFeature([Ingredient, Recipe])],
```

Comment:
I don't know if it's super important in the long run, but it would be useful to me to know what the relationships of the "imports", "providers", and "exports" arrays, and how they are related across modules.

Reply:
Good idea - I added some info to `NestJS.md` on that